### PR TITLE
Derelict Cyborg event

### PR DESCRIPTION
**Pull request by GoldenCan**
## About the PR
Added "Derelict Cyborg" as a midround spawn event.
(It now also fixes https://github.com/space-wizards/space-station-14/issues/31485, the commit which fixes it is https://github.com/space-wizards/space-station-14/pull/32499/commits/4b633fde9c84778e2817025b1d38bd2549202c52)

Lore/in-universe explanation (this exact text does not appear in-game): "_A cyborg that, through some mishap, drifted away from its station or shuttle. Perhaps its station was destroyed? After years a drift in space the cyborgs enters the proximity of a particular space station..._"

The Derelict Cyborg spawns in space near the station, with a large battery so the can last a while (roundstart borgs have medium batteries, Syndicate Assault borgs have hyper batteries.) and several modules that can help them board the station: A GPS module so the Derelict Cyborg can find the station even if it is too far away to see, a Fire Extinguisher module so the Derelict can reach the station, and a basic Tool module for simple general purpose use and so it can crowbar its way onto the station if the station is without power.
(If the Derelict Cyborg accidentaly uses it entire fire extinguisher without reaching the station (which i find unlikely as the station can be reached in about 1 use) the Derelict Cyborg still has access to the Commons and Binary (but not Science) radio channels, which it can, along with its GPS module, use to attempt to get someone to help them out of its situation.)

## About the PR: Randomized Lawset
The Derelict Cyborg has a randomized lawset (although this has not been added to this PR yet, but it does work) which is the result of hitting it several times with the same system which is used for the Ion Storm event. This mean that the Derelict Cyborg will likely neither be crew-aligned nor aligned against the crew, but rather have their own strange rules to follow. Although there is always the chance that it will be a killer robot or basically just crew-aligned - it is all randomized.

## Differences and similarities with NT Cyborgs and Syndicate Cyborgs.
The Derelict Cyborg does not appear on the Robotics console - just like syndicate cyborgs don't appear on it.
The Derelict Cyborg has access to Commons and Binary radio channels, but (as of now) not the Science channel.
The Derelict Cyborg DOES have all-access, as i believe it would otherwise have a hard time navigating the station. _And it was a regular NT cyborg at one point, after all.
The Derelict Cyborg has a large battery, which lasts longer than the generic Nanotrasen cyborg's medium battery, but not as long as the Syndicate Assault Borg's hyper battery. The Derelict Cyborg will likely find a place to recharge long before it needs to. (Unless it is been played on Reach - i don't think it can recharge there without having its battery removed.)
The Derelict Cyborg is often neither actively hostile nor entirely friendly.

## Why

Ion storm laws can be fun to interact with, but they have to be counterbalanced with the fact that they typically negatively affect the crew's Cyborgs. The more common ion storms are the less desirable cyborgs will become.
The Derelict Cyborg allows players (both the ones playing the Derelict Cyborg and the players they interact with) to experience the interesting effects of severely ion-storm affected laws without interfereing with the station's existing cyborgs.

The Derelict Cyborg would have fun and interesting roleplay and gameplay possibilites with its randomized lawset, which also has the potential for a great amount of variety.

## Balance
**IMPACT ON THE ROUND/POTENTIAL THREAT**
The Derelict Cyborg is often neither actively hostile nor entirely friendly.
The Derelict Cyborg does not have any particularly damaging tools at its disposal, and as such won't be a great threat even if it is hostile - although it does have a fire extinguisher which can be harmful to Slimepeople.
The Derelict Cyborg cannot repair itself.

Generally, the Derelict Cyborg will have less of an impact on the round compared to actual midround antagonists, and it does not have the same potential to become a threat against Crew or Antagonists as a Closet Skeleton or Rat King has. The Derelict Cyborg's strength can then be said to come moreso from interesting and varied gameplay and roleplay implecations.

## Technical details

The Derelict Cyborg uses a new component: StartIonStormedComponent.

**cs changes:**
IonStormRule.cs previously checked every cyborg and then had a chance to alter the cyborg's laws. The code for subjecting an individual cyborg has now been moved to IonStormSystem.cs. IonStormRule.cs now uses IonStormSystem.
Added StartIonStormedSystem.cs which subjects an individual cyborg to the results of an ion storm an amount of times as specified by StartIonStormedComponent. StartIonStormedSystem.cs uses IonStormSystem.cs.

**Derelict Cyborg itself:**
**The event**
DerelictCyborgSpawn is a new station event listed under BasicAntagEventsTable (the same table as both most event antagonists as well the Rat King and Closet Skeleton, two free agents.).
The event has a weight of 8, which is slightly higher than that of the closet skeleton. I decided this weight because the Derelict Cyborg has less potential for disruption than the closet skeleton (or any of the antagonists under BasicAntagEventsTable), due to both its laws and as it is limited to its starting modules (unless a non-silicon gives it more modules for some reason) as explained under "balance".

The Derelict Cyborg uses SpaceSpawnRule with a spawnDistance of 0, exactly like the space dragon, to make it spawn in space but close to the station.

## Media

Derelict Cyborgs ingame:
zDerelictDone3
Derelict Cyborg modules, etc.:
zDerelict2
Derelict Cyborg ghostrole event:
(Sometimes you spawn touching the station, other times you need to use your GPS and fire extinguisher once to get close to it.)
zDerelictSometimes you spawn touching the station
Derelict Cyborg once spawned (it spawns close enough to the station that it can reach it in about 1 use of its fire extinguisher):
zDerelictDone2
Two examples randomized ion-storm lawset that a Derelict Cyborg can spawn with (first picture is a bit outdated.):
zDerelictLaws2
zDerelictSomeLawsAreJustWeird

**DERELICT CYBORG NO LONGER APPEARS ON THE ENDROUND "GAME INFORMATION" SCREEN**
It of course still appears on the endround "Player Manifest" screen. The change was made as the Derelict Cyborg was not an antagonist and only antagonists appeared on that screen. Additionaly, other silicons with altered laws - like emagged ones or ones hit by ion storms - do not show up on the "Game Information" screen either.
zDerelictEndroundBlank

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

**Changelog**

:cl: ScarKy0, GoldenCan
- add: The Derelict Cyborg - a broken cyborg with altered laws due to year of exposure to ion storms - can now appear as a ghost role through a new midround event.
- fix: An ion storm affecting a Cyborg can no longer alter the law order of the AI, cyborgs created later in the round or AI's and cyborgs in subsequent rounds.

### CODE REVIEW

**Comment by KaiserMaus** [comment]
Derelictcyborgteor he he or syndrome derelict

**Comment by ArcaneOcto** [comment]
YES

**Comment by WhiteGorila** [comment]
QUICK REMOVE THE WORD ANTAGONIST FROM THE PR SO IT GETS MERGED

**Comment by GoldenCan** [comment]
> QUICK REMOVE THE WORD ANTAGONIST FROM THE PR SO IT GETS MERGED

Done.

**Comment by ArtisticRoomba** [comment]
_I've been in hibernation for 84 years..._

Can't wait to see the role play content out of this :)

File: Content.Server/Silicons/Laws/IonStormSystem.cs [hunk 1]
```diff
@@ -0,0 +1,283 @@
+using System.Linq;
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/IonStormSystem.cs:1]:
```suggestion
```
keep in alphabetic order

File: Content.Server/Silicons/Laws/IonStormSystem.cs [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+using Content.Shared.Dataset;
+using Content.Shared.FixedPoint;
+using Content.Shared.GameTicking.Components;
+using Content.Shared.Random;
+using Content.Shared.Random.Helpers;
+using Content.Shared.Silicons.Laws;
+using Content.Shared.Silicons.Laws.Components;
+using Content.Shared.Station.Components;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/IonStormSystem.cs:14]:
```suggestion
using Robust.Shared.Random;
using System.Linq;
```
here

File: Content.Server/Silicons/Laws/IonStormSystem.cs [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    [ValidatePrototypeId<DatasetPrototype>]
+    private const string Concepts = "IonStormConcepts";
+    [ValidatePrototypeId<DatasetPrototype>]
+    private const string Drinks = "IonStormDrinks";
+    [ValidatePrototypeId<DatasetPrototype>]
+    private const string Foods = "IonStormFoods";
+
+    public void IonStormTarget(EntityUid ent, SiliconLawBoundComponent lawBound, TransformComponent xform, IonStormTargetComponent target, EntityUid? chosenStation, bool ignoreStation = false, bool DoNotAdminlog = false)
+    {
+            // only affect law holders on the station unless ignoreStation is true.
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/IonStormSystem.cs:65]: Fix the intendation throughout this whole function.
My IDE is screaming at me.

File: Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+using Content.Shared.Random;
+using Robust.Shared.Prototypes;
+
```
Comment (slarticodefast) [inline Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs:3]:
```suggestion
```
not needed

File: Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+namespace Content.Shared.Silicons.Laws.Components;
+
+/// <summary>
+/// Runs the IonStormSystem on an entity IonStormAmount times.
```
Comment (slarticodefast) [inline Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs:7]:
```suggestion
/// Applies law changing ion storms on an entity IonStormAmount times when the enitity is spawned.
```

File: Content.Server/StationEvents/Events/IonStormRule.cs [hunk 6]
```diff
@@ -1,64 +1,15 @@
 using System.Linq;
```
Comment (slarticodefast) [inline Content.Server/StationEvents/Events/IonStormRule.cs:1]:
```suggestion
```
not needed

File: Content.Server/StationEvents/Events/IonStormRule.cs [hunk 7]
```diff
@@ -1,64 +1,15 @@
...
 using Content.Server.Silicons.Laws;
```
Comment (slarticodefast) [inline Content.Server/StationEvents/Events/IonStormRule.cs:2]:
```suggestion
```

File: Resources/Prototypes/Entities/Mobs/Cyborgs/base_borg_chassis.yml [hunk 8]
```diff
@@ -317,3 +317,34 @@
...
+    factions:
+    - NanoTrasen #The seemingly best fit. It was a regular NT cyborg once, after all.
+  - type: Access
+    enabled: false
+    groups:
+    - AllAccess #Randomized access would be fun. AllAccess is the best i can think of right now that does make it too hard for it to enter the station or navigate it..
+  - type: AccessReader
+    access: [["Command"], ["Research"]]
+  - type: SiliconLawProvider
+    laws: Crewsimov #Although this will be randomized.
```
Comment (slarticodefast) [inline Resources/Prototypes/Entities/Mobs/Cyborgs/base_borg_chassis.yml:336]:
```suggestion
```
already inherited

File: Resources/Prototypes/Entities/Mobs/Cyborgs/base_borg_chassis.yml [hunk 9]
```diff
@@ -317,3 +317,34 @@
...
+  - type: IntrinsicRadioTransmitter
+    channels:
+    - Binary
+    - Common
+    - Science
+  - type: ActiveRadio
+    channels:
+    - Binary
+    - Common
+    - Science
```
Comment (slarticodefast) [inline Resources/Prototypes/Entities/Mobs/Cyborgs/base_borg_chassis.yml:346]:
```suggestion
```
same

File: Resources/Prototypes/GameRules/events.yml [hunk 10]
```diff
@@ -545,3 +546,28 @@
...
+  - type: SpaceSpawnRule
+    spawnDistance: 0
+  - type: AntagSpawner
+    prototype: PlayerBorgDerelict
+  - type: AntagSelection
+    agentName: derelict-cyborg-round-end-agent-name
+    definitions:
+#      briefing:
+#        text: derelict-cyborg-role-greetin
+#        color: Blue
```
Comment (slarticodefast) [inline Resources/Prototypes/GameRules/events.yml:569]: What's up with this?
Also not sure if the event should count as an antagonist, but I will ask the admins

File: Resources/Prototypes/GameRules/events.yml [hunk 11]
```diff
@@ -545,3 +546,28 @@
     maxOccurrences: 1 # this event has diminishing returns on interesting-ness, so we cap it
     weight: 5
   - type: MobReplacementRule
+
+- type: entity
+  parent: BaseGameRule
+  id: DerelictCyborgSpawn
+  components:
+  - type: StationEvent
+    weight: 8
```
Comment (slarticodefast) [inline Resources/Prototypes/GameRules/events.yml:555]:
```suggestion
    weight: 5
```
I would say make it a little rarer, same as closet skeletons. Given that they have all access they will probably cause quite a lot of chaos.

File: Content.Server/Silicons/Laws/StartIonStormedSystem.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
+using Content.Shared.Silicons.Laws.Components;
+using Content.Shared.Administration.Logs;
+using Content.Shared.Database;
+using Content.Shared.Silicons.Laws;
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/StartIonStormedSystem.cs:4]:
```suggestion
```
not needed

File: Content.Server/Silicons/Laws/StartIonStormedSystem.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    [Dependency] private readonly ISharedAdminLogManager _adminLogger = default!;
+    [Dependency] private readonly SiliconLawSystem _siliconLaw = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<StartIonStormedComponent, MapInitEvent>(OnMapInit);
+    }
+
+    private void OnMapInit(EntityUid uid, StartIonStormedComponent component, ref MapInitEvent args)
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/StartIonStormedSystem.cs:23]:
```suggestion
    private void OnMapInit(Entity<StartIonStormedComponent> ent, ref MapInitEvent args)
```
We always use Entity<Component> when possible.
You can get the uid via `ent.Owner`
and the component via `ent.Comp`

File: Content.Server/Silicons/Laws/StartIonStormedSystem.cs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        if (!TryComp<SiliconLawBoundComponent>(uid, out var lawBound))
+            return;
+        if (!TryComp<TransformComponent>(uid, out var xform))
+            return;
+        if (!TryComp<IonStormTargetComponent>(uid, out var target))
+            return;
+
+        for (int currentIonStorm = 1; currentIonStorm <= component.IonStormAmount; currentIonStorm++)
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/StartIonStormedSystem.cs:32]:
```suggestion
        for (int currentIonStorm = 0; currentIonStorm < component.IonStormAmount; currentIonStorm++)
```
Programmers start counting at 0 😄 
(your way is fine of course)

File: Resources/Prototypes/Entities/Mobs/Player/silicon.yml [hunk 15]
```diff
@@ -515,3 +515,52 @@
...
+  components:
+  - type: ContainerFill
+    containers:
+      borg_brain:
+        - MMIFilled
+  - type: ItemSlots
+    slots:
+      cell_slot:
+        name: power-cell-slot-component-slot-name-default
+        startingItem: PowerCellHigh
```
Comment (slarticodefast) [inline Resources/Prototypes/Entities/Mobs/Player/silicon.yml:553]: This one isn't really needed, for the normal borg it exists mostly for debugging purposes.
```suggestion
```

File: Resources/Prototypes/Entities/Mobs/Player/silicon.yml [hunk 16]
```diff
@@ -515,3 +515,52 @@
...
+
+- type: entity
+  id: PlayerBorgDerelictGhostRole
+  parent: PlayerBorgDerelictBattery
```
Comment (slarticodefast) [inline Resources/Prototypes/Entities/Mobs/Player/silicon.yml:557]:
```suggestion
  parent: PlayerBorgDerelict
```

**Review by slarticodefast — CHANGES_REQUESTED** [review]
Needs some cleanup, but I love the idea.

File: Resources/Textures/Mobs/Silicon/chassis.rsi/meta.json [hunk 17]
```diff
@@ -1,5 +1,5 @@
 {
-  "version": 1,
+  "version": 2,
```
[the remark this answers is not in the document]
Comment (GoldenCan) [reply Resources/Textures/Mobs/Silicon/chassis.rsi/meta.json:2]: oops

**Comment by slarticodefast** [comment]
Holy web edit. RIP my inbox.
Please make these changes locally and test them in-game to make sure they work as intended. Web edits are prone to errors and frequently add wrong whitespace.

**Comment by GoldenCan** [comment]
> Holy web edit. RIP my inbox. Please make these changes locally and test them in-game to make sure they work as intended. Web edits are prone to errors and frequently add wrong whitespace.

I will test these locally, and i will make any further changes there.

File: Content.Server/Silicons/Laws/StartIonStormedSystem.cs (hunk 13 above)
Comment (GoldenCan) [reply Content.Server/Silicons/Laws/StartIonStormedSystem.cs]: I tried to make a change here - i hope i understood this correctly.

File: Content.Server/StationEvents/Events/IonStormRule.cs (hunk 7 above)
Comment (GoldenCan) [reply Content.Server/StationEvents/Events/IonStormRule.cs:1]: I don't understand this change, it doesn't build if 'using Content.Server.Silicons.Laws;' is removed.

**Comment by superjj18** [comment]
Command/Sci on their way to instantly reborg every derelict Borg on sight to ensure no fun is had

**Comment by a-person5660** [comment]
if your not intentionally letting malf borgs live then are you really playing science? :trollface:

**Comment by GoldenCan** [comment]
A fix for https://github.com/space-wizards/space-station-14/issues/31485 has been included as a part of the PR's alteration of the code for IonStorms.

**Comment by GoldenCan** [comment]
> A fix for #31485 has been included as a part of the PR's alteration of the code for IonStorms.

The change to fix this was really simple:
https://github.com/space-wizards/space-station-14/pull/32499/commits/4b633fde9c84778e2817025b1d38bd2549202c52

**Comment by slarticodefast** [comment]
> The change to fix this was really simple:

Heroic fix, this one was causing a lot of admin problems, lol
I will try and give the PR another review later today so we can get it in before PJB's borg changes, so you don't have to fix the massive merge conflict from that

**Comment by ScarKy0** [comment]
>I will try and give the PR another review later today so we can get it in before PJB's borg changes, so you don't have to fix the massive merge conflict from that

Be blessed I was scared thats gonna be the case

**Comment by GoldenCan** [comment]
The Derelict Cyborg no longer appear on the "Game Information" screen shown at the end of the round.
It of course still appears on the endround "Player Manifest" screen. The change was made to add consistency as the Derelict Cyborg was not an antagonist and only antagonists appeared on that screen. Additionaly, other silicons with altered laws - like emagged ones or ones hit by ion storms - do not show up on the "Game Information" screen either.

File: Content.Server/Silicons/Laws/IonStormSystem.cs [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+    [ValidatePrototypeId<DatasetPrototype>]
+    private const string AllergySeverities = "IonStormAllergySeverities";
+    [ValidatePrototypeId<DatasetPrototype>]
+    private const string Concepts = "IonStormConcepts";
+    [ValidatePrototypeId<DatasetPrototype>]
+    private const string Drinks = "IonStormDrinks";
+    [ValidatePrototypeId<DatasetPrototype>]
+    private const string Foods = "IonStormFoods";
+
+    //Randomly alters the laws of an individual silicon.
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/IonStormSystem.cs:62]:
```suggestion
    /// <summary>
    /// Randomly alters the laws of an individual silicon.
    /// </summary>
```
This way the summary shows up in the IDE if you hover over the function with your mouse.

File: Content.Server/Silicons/Laws/IonStormSystem.cs [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+    public void IonStormTarget(Entity<SiliconLawBoundComponent, IonStormTargetComponent> ent, TransformComponent xform, EntityUid? chosenStation, bool adminlog = true)
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/IonStormSystem.cs:63]:
```suggestion
    public void IonStormTarget(Entity<SiliconLawBoundComponent, IonStormTargetComponent> ent, bool adminlog = true)
```
not used

File: Content.Server/Silicons/Laws/StartIonStormedSystem.cs [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+        SubscribeLocalEvent<StartIonStormedComponent, MapInitEvent>(OnMapInit);
+    }
+
+    //private void OnMapInit(EntityUid uid, StartIonStormedComponent component, ref MapInitEvent args)'
+    private void OnMapInit(Entity<StartIonStormedComponent> ent, ref MapInitEvent args)
+    {
+        if (!TryComp<SiliconLawBoundComponent>(ent.Owner, out var lawBound))
+            return;
+        if (!TryComp<TransformComponent>(ent.Owner, out var xform))
+            return;
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/StartIonStormedSystem.cs:28]:
```suggestion
```
not needed

File: Content.Server/Silicons/Laws/StartIonStormedSystem.cs [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+        if (!TryComp<IonStormTargetComponent>(ent.Owner, out var target))
+            return;
+
+        for (int currentIonStorm = 0; currentIonStorm < ent.Comp.IonStormAmount; currentIonStorm++)
+        {
+            _ionStorm.IonStormTarget((ent.Owner, lawBound, target), xform, null, false);
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/StartIonStormedSystem.cs:34]:
```suggestion
            _ionStorm.IonStormTarget((ent.Owner, lawBound, target), false);
```

File: Content.Server/StationEvents/Events/IonStormRule.cs (hunk 7 above)
Comment (slarticodefast) [reply Content.Server/StationEvents/Events/IonStormRule.cs:1]: Oops, my bad. VsCodium sometimes lies to me about unnecessary imports.

File: Content.Server/StationEvents/Events/IonStormRule.cs [hunk 22]
```diff
@@ -74,217 +24,7 @@ protected override void Started(EntityUid uid, IonStormRuleComponent comp, GameR
...
-                }
-            }
-
-            _adminLogger.Add(LogType.Mind, LogImpact.High, $"{ToPrettyString(ent):silicon} had its laws changed by an ion storm to {laws.LoggingString()}");
-
-            // laws unique to this silicon, dont use station laws anymore
-            EnsureComp<SiliconLawProviderComponent>(ent);
-            var ev = new IonStormLawsEvent(laws);
-            RaiseLocalEvent(ent, ref ev);
+            _ionStorm.IonStormTarget((ent, lawBound, target), xform, chosenStation);
```
Comment (slarticodefast) [inline Content.Server/StationEvents/Events/IonStormRule.cs:27]:
```suggestion
            _ionStorm.IonStormTarget((ent, lawBound, target));
```

File: Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs [hunk 23]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (slarticodefast) [inline Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs:1]:
```suggestion
```

File: Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+/// </summary>
+[RegisterComponent]
+public sealed partial class StartIonStormedComponent : Component
+{
+    /// <summary>
+    /// Amount of times that the ion storm will be run on the entity on spawn.
+    /// </summary>
+    [DataField]
+    public int IonStormAmount = 1;
+}
```
Comment (slarticodefast) [inline Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs:15]: Are you using an IDE or something like notepad? Our .editorconfig is set up to add a new line at the end of each file, so I wonder why there is one missing here.

File: Resources/Locale/en-US/silicons/derelict-cyborg-role.ftl [hunk 25]
```diff
@@ -0,0 +1 @@
+derelict-cyborg-round-end-agent-name = derelict cyborg
```
Comment (slarticodefast) [inline Resources/Locale/en-US/silicons/derelict-cyborg-role.ftl:1]: remove this file

File: Resources/Prototypes/Entities/Mobs/Player/silicon.yml [hunk 26]
```diff
@@ -515,3 +515,37 @@
...
+      cell_slot:
+        name: power-cell-slot-component-slot-name-default
+        startingItem: PowerCellHigh
+  - type: RandomMetadata
+    nameSegments: [names_borg]
+
+- type: entity
+  id: PlayerBorgDerelictGhostRole
+  parent: PlayerBorgDerelict
+  suffix: Battery, Ghost role
```
Comment (slarticodefast) [inline Resources/Prototypes/Entities/Mobs/Player/silicon.yml:543]:
```suggestion
  suffix: Ghost role
```

**Review by slarticodefast — CHANGES_REQUESTED** [review]

File: Resources/Locale/en-US/silicons/derelict-cyborg-role.ftl (hunk 25 above)
Comment (GoldenCan) [reply Resources/Locale/en-US/silicons/derelict-cyborg-role.ftl:1]: i thought i had removed that already. strange.

Comment (GoldenCan) [reply Resources/Locale/en-US/silicons/derelict-cyborg-role.ftl:1]: the file does not appear on my system.

Comment (GoldenCan) [reply Resources/Locale/en-US/silicons/derelict-cyborg-role.ftl]: I think i forgot to push that.

File: Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs (hunk 24 above)
Comment (GoldenCan) [reply Content.Shared/Silicons/Laws/Components/StartIonStormedComponent.cs:14]: For .cs i used Visual Studio 2022.

**Comment by slarticodefast** [comment]
Ok, the code looks good to me, but there is one previously existing bug that will need to be fixed first so that this does not cause unneccessary problems for admins:
Normally when a borg gets ion stormed they get labeled as an antagonist in the admin overlay. However, an empty chassis can get its laws changed as well and the antag label will be missing if a player takes control at a later point.
This is especially problematic with the derelict borg, since it always gets spawned and ion stormed before the player takes control.

Derelict borg:
grafik
Normal borg that gets ion stormed:
grafik

**Comment by GoldenCan** [comment]
I see. I don't think that is a huge problem for the derelict borg as it has a unique appearance and always has altered laws. I can give it a try.

**Comment by GoldenCan** [comment]
> I see. I don't think that is a huge problem for the derelict borg as it has a unique appearance and always has altered laws. I can give it a try.

(I don't know what i am doing.)

**Comment by slarticodefast** [comment]
I can give you a ping on Discord later when I'm home again. I got an idea in mind how to fix it, but it will need some testing.

**Comment by GoldenCan** [comment]
> I can give you a ping on Discord later when I'm home again. I got an idea in mind how to fix it, but it will need some testing.

I am currently trying something out which i think will work.

**Comment by GoldenCan** [comment]
> > I can give you a ping on Discord later when I'm home again. I got an idea in mind how to fix it, but it will need some testing.
> 
> I am currently trying something out which i think will work.

Although it would only fix the issue for the Derelict Cyborg. I also won't be able to work on it right now.

**Comment by GoldenCan** [comment]
I fixed it for cyborgs with the StartIonStormedComponent (which is just the Derelict Cyborg) by copying what emagged cyborgs do. It works for Derelict borgs spawn in throught the random event, and it also functions correctly if you removed the positronic brain and insert a new one manually.
zDerelictThisWorkIguess

**Review by slarticodefast — CHANGES_REQUESTED** [review]
Ok, I don't like how this copy pastes the stuff from emagged borgs and does not solve the underlying problem for ion storms. This will need some cleanup in general to unify the emag and ion storm mind role changes:
- EnsureEmaggedRole should probably be renamed to EnsureSubvertedSiliconRole since it is also used for ion storms (why the hell does it check for component.AntagonistRole, but then doesn't even use it? On the current master at least, so make sure to merge it first). Just remove the EmagSiliconLawComponent from the parameters and the datafield from the component, since it is not used.
- OnIonStormLaws should not check for EmagSiliconLawComponent when adding the role (this seemed to be a bandaid to get the role to add to the mind, but it wasn't even used like I mentioned above), because if someone makes a borg that can be ion stormed, but not emagged they would not get the MindRoleSubvertedSilicon role and thus would not show up as a traitor to admins.
- simply add a boolean "Subverted" datafield to SiliconLawProviderComponent
- set that to true when ion storming or emagging
- in a subscription to MindAddedMessage with the SiliconLawProviderComponent, if that datafield is true, call EnsureSubvertedSiliconRole, do the reverse and remove the role for the MindRemovedMessage 
- remove the OnEmagMindAdded, OnemagMindRemoved, OnStartIonstormMindAdded, OnstartIonStormedMindRemoved since we just unified those

**Comment by GoldenCan** [comment]
> Ok, I don't like how this copy pastes the stuff from emagged borgs and does not solve the underlying problem for ion storms. This will need some cleanup in general to unify the emag and ion storm mind role changes:

I know that and i knew that...
It was a temporary solution, but i ran low on motivation, which i heavily needing due to my lack of experience which would have made fixing it properly uphill - _but probably not impossible_ - battle.

**Comment by slarticodefast** [comment]
Good news is I just approved Scar's Ion storm refactor which splits off that part from your PR and addresses the review. So this PR will then only need the new cyborg and the StartIonStormedComponent/System,

**Comment by slarticodefast** [comment]
Some information about the status of this PR;
- #33311 accidentally got merged instead of squash merged
- It was based off this PR, but reverted the derelict borg part and kept the ion storm changes
- so technically this PR was merged and then reverted, which is why GitHub shows it as merged, but the derelict borg is currently *not* in the game
- ScarKy0 will make a new PR adding the derelict borg again

**Pull request merged**